### PR TITLE
Yury_Vislobodsky Module 7: Java and XML programming

### PR DESCRIPTION
**Pull request by YurasicV**

### CODE REVIEW

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/resources/pom.xml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/resources/pom.xml]: What about formatting? Just press Ctrl+Alt+L in IDEA...

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/exception/PlayParserException.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package javase.t01.exception;
+
+/**
+ * Common exception raising when some troubles take place with Play Parser
+ * Created by Yury Vislobodsky on 10.04.2016.
+ */
+public class PlayParserException extends Exception {
+    private final Exception exception;
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/exception/PlayParserException.java]: It's wrong. Just use super(Throwable) constructor to keep information about the cause of the exception.
Your implementation won't include the stacktrace and it's bad for debugging.

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/dom/PlayParserDom.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import java.io.File;
+import java.io.IOException;
+import java.io.StringReader;
+
+/**
+ * Play Parser Class (method DOM)
+ * Created by Yury Vislobodsky on 06.04.2016.
+ */
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/dom/PlayParserDom.java]: I would prefer make an interface PlayParser:

```
public interface PlayParser {
         Play parse(String fileName) throws PlayParserException;
}
```

and create several implementations like PlayDomParser, PlaySaxParser and PlayStaxParser.


Use OOP power :-)

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/TagNames.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package javase.t01.play;
+
+/**
+ * Tag names enums for Play Parsers
+ * Created by Yury Vislobodsky on 06.04.2016.
+ */
+public enum TagNames {
+    SPEECH, SPEAKER, LINE, STAGEDIR, OTHER;
+
+    public static TagNames getElementTagName(String element) {
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/TagNames.java]: Use TagName.valueOf(String).

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Totals.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * Totals class for Play
+ * Created by Yury Vislobodsky on 06.04.2016.
+ */
+public class Totals {
+    private int speechCount;
+    private int wordsCount;
+
+    public Totals() {
+        speechCount = 0;
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Totals.java]: It's redundant.
https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html
Take a look at "Default Values" section.

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Play.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * Class for play (map of speeches)
+ * Created by Yury Vislobodsky on 06.04.2016.
+ */
+public class Play {
+    private Map<String, Totals> list;
+
+    public Play() {
+        list = new TreeMap<>();
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Play.java]: You can move it to line 11.

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Play.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    public void add(Speech speech) {
+        if (speech == null) {
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Play.java]: It confuses an API user. Just remove this section and allow NullPointerException to be thrown to detect the error earlier.

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Speech.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ * Speech class for Play
+ * Created by Yury Vislobodsky on 06.04.2016.
+ */
+public class Speech {
+    private List<String> speakers;
+    private int wordsCount;
+
+    public Speech() {
+        speakers = new ArrayList<>();
+        wordsCount = 0;
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Speech.java]: It's redundant,
https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html
Take a look at "Default Values" section.

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Speech.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * Speech class for Play
+ * Created by Yury Vislobodsky on 06.04.2016.
+ */
+public class Speech {
+    private List<String> speakers;
+    private int wordsCount;
+
+    public Speech() {
+        speakers = new ArrayList<>();
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Speech.java]: Can be moved to line 11.

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Speech.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        wordsCount = 0;
+    }
+
+    public void addSpeaker(String speaker) {
+        speakers.add(speaker);
+    }
+
+    public void addLine(String line) {
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Speech.java]: Incorrect method name. It does not do what it claims.

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Speech.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        this.wordsCount += line.split("[^A-Za-z]+").length;
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Speech.java]: Consider using a compiled regexp to count words. String.split compiles a new regexp per method call and it's quite expensive operation, it also produces redundant split words.

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t02/CreatePomDemo.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        Element plugin = appendChildElement(document, plugins, tagPlugin, "");
+        appendChildElement(document, plugin, tagGroupId, "org.apache.maven.plugins");
+        appendChildElement(document, plugin, tagArtifactId, "maven-compiler-plugin");
+        Element configuration = appendChildElement(document, plugin, "configuration", "");
+        appendChildElement(document, configuration, "source", "1.7");
+        appendChildElement(document, configuration, "target", "1.7");
+
+        Element dependencies = appendChildElement(document, project, "dependencies", "");
+
+        Element dependency = appendChildElement(document, dependencies, tagDependency, "");
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t02/CreatePomDemo.java]: You can extract the method createDependencyNode(...), don't repeat yourself.

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t02/CreatePomDemo.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+/**
+ * Create pom.xml using DOM Parser (Demo)
+ * Created by Yury Vislobodsky on 10.04.2016.
+ */
+public class CreatePomDemo {
+    private CreatePomDemo() {}
+
+    public static void main(String[] args) throws ParserConfigurationException, IOException, TransformerException {
+        final String tagGroupId = "groupId";
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t02/CreatePomDemo.java]: Extract as constants.

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Play.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
+package javase.t01.play;
+
+import java.util.Map;
+import java.util.TreeMap;
+
+/**
+ * Class for play (map of speeches)
```
Comment (technicaltaskjava) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/java/javase/t01/play/Play.java]: Useless javadoc.

File: Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/resources/pom.xml (hunk 1 above)
Comment (YurasicV) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/resources/pom.xml]: This POM file was generated by my java app according to condition of Task 2. 
It was formatted by method transformer.setOutputProperty(OutputKeys.INDENT, "yes");
How may I do it better? Not in IDE, but programmatically :)

Comment (YurasicV) [inline Yuri_Vislobodsky/Module 7 - Java and XML programming/javase07/src/main/resources/pom.xml]: Hello,
thanks for helpful advices.
I am sure I will consider them in the future.

Just a moment:
This POM file was generated by my java app according to condition of 
Task 2.
It was formatted by method 
transformer.setOutputProperty(OutputKeys.INDENT, "yes");
How may I do it better? Not in IDE, but programmatically :)

Yury

14.04.2016 17:06, technicaltaskjava пишет:

**Pull request merged**